src/Components/Firestore.js

```diff
@@ -211,35 +208,105 @@ export function generateId() {
+  }
+}
+
 // Handle "watchlistData / reactions" on firestore
 
-export async function GetListReactions(docId, setListRxns) {
-  try {
-    let animeRef = doc(db, "watchlistData", docId, "reactions", "emojis");
-    let querySnapshot = await getDoc(animeRef);
+export async function getUserReactions(uid, docId) {
+  let reactionsRef = doc(db, "users", uid, "reactions", docId);
+  let reactionSnap = await getDoc(reactionsRef);
+  if (!reactionSnap.data()) {
     // Use default value if the document doesn't exist yet.
-    if (!querySnapshot.data()) {
-      setListRxns({
-        emojis: { applause: [], heart: [], trash: [] },
-      });
-      return;
-    }
-    let temp = querySnapshot.data();
-    setListRxns(temp);
-  } catch (error) {
-    console.error("Error loading data from Firebase Database", error);
+    return {
+      uniqueEntityId: docId,
+      applause: false,
+      heart: false,
+      trash: false,
+    };
+  } else {
+    return reactionSnap.data();
   }
 }
 
-export async function SaveListReactionsToFirestore(docId, updatedRxns) {
-  let Ref = doc(db, "watchlistData", docId, "reactions", "emojis");
+export async function getReactionCount(uid, docId) {
+  let rxnCountRef = doc(db, "watchlistData", docId);
+  let rxnCountSnap = await getDoc(rxnCountRef);
+  return rxnCountSnap.data();
+}
+
+export async function SaveReactionsToFirestore(
+  uid,
+  docId,
+  updatedRxns,
+  reactionTo,
+  reaction,
+  commentOwnerId
+) {
+  const value = updatedRxns[reaction] ? 1 : -1;
+  let key;
+  if (reaction === "trash") key = "trashCount";
+  else if (reaction === "heart") key = "heartCount";
+  else if (reaction === "applause") key = "applauseCount";
+
+  const uniqueEntityId =
+    reactionTo === "comments" ? docId + commentOwnerId : docId;
+
+  const rxnRef = doc(db, "users", uid, "reactions", uniqueEntityId);
+  let rxnCountRef;
+
+  if (reactionTo === "list") {
+    rxnCountRef = doc(db, "watchlistData", docId);
+  } else if (reactionTo === "reviews") {
+    rxnCountRef = doc(db, "animeData", docId, "reviews", uid);
+  } else if (reactionTo === "comments") {
+    rxnCountRef = doc(db, "watchlistData", docId, "reviews", commentOwnerId);
+  }
   try {
-    await setDoc(Ref, updatedRxns, { merge: true });
+    await Promise.all(
+      [
+        setDoc(
+          rxnRef,
+          { [reaction]: updatedRxns[reaction], uniqueEntityId },
+          { merge: true }
+        ),
+      ],
+      updateDoc(rxnCountRef, { [key]: increment(value) })
+    );
   } catch (error) {
     console.error("Error writing review data to Firestore", error);
   }
 }
 
+// Clears all stored emoji selections for users who reacted to this review/comment
+export async function deleteAllReactions(uniqueEntityId) {
+  const allRxns = query(
+    collectionGroup(db, "reactions"),
+    where("uniqueEntityId", "==", uniqueEntityId)
+  );
+  try {
+    const querySnapshot = await getDocs(allRxns);
+    querySnapshot.forEach((doc) => {
+      deleteDoc(doc.ref);
```

Comment:
This will require one user (the user doing the deletion) to be able to delete documents from a bunch of other users' `/users/:uid/reactions` collections.  I know we discussed an access-control principal that the owning user is the only one able to write to their `/users/:uid` document and subcollections, and so these deletions would be disallowed if we add such a rule.  

I'm also worried this could result in a lot of deletions all happening from one user's web app, in the case of a review that had a lot of reactions being deleted.  Is there any harm in leaving these zombie reactions in the respective users' `reactions` collections?  Maybe at least util a server-side vacuum job can clean them up?

Reply:
All very fair concerns!

The unwanted behavior from zombie reactions, is if someone updates a popular review/comment by deleting it and then reposting (instead of editing it - the way god intended) then all of their friends who had interacted with the old review/comment will still have those emojis marked as selected, despite the emoji count of this new review/comment being 0. 

Admittedly, this is probably an unlikely case, but I did run into it while testing things out so take that for what you will 🙂.  I'm satisfied it we just run a server-side sweeper daily or so.. 

Lmkwyt...but for now I'll delete this `deleteAllReactions` function.